src/main/java/org/cryptomator/cloudaccess/MetadataRequestAggregator.java

```diff
@@ -70,22 +74,24 @@ public MetadataCachingProviderDecorator(CloudProvider delegate, Duration cacheEn
 
 	@Override
 	public CompletionStage<CloudItemList> list(CloudPath folder, Optional<String> pageToken) {
-		return delegate.list(folder, pageToken) //
-				.whenComplete((cloudItemList, exception) -> {
-					evictIncludingDescendants(folder);
-					if (exception == null) {
-						assert cloudItemList != null;
-						cloudItemList.getItems().forEach(metadata -> itemMetadataCache.put(metadata.getPath(), CompletableFuture.completedFuture(metadata)));
-					}
-				});
+		try {
+			return cloudItemListRequestAggregator.get(Map.entry(folder, pageToken), () -> delegate.list(folder, pageToken).whenComplete((metadata, throwable) -> {
+				// immediately invalidate cache in case of exceptions (except for NOT FOUND):
+				if (throwable != null && !(throwable instanceof NotFoundException)) {
+					invalidateAggregators(folder);
+				}
+			}));
+		} catch (ExecutionException e) {
+			return CompletableFuture.failedFuture(e);
+		}
```

Comment:
1. use `cache.getUnchecked()` to avoid try-catch (only if we know for sure that the loading method doesn't throw checked exceptions)
2. instead of two nested lambdas, maybe the outer lambda can be a `private CompletionStage<CloudItemList> listUncached(folder, pageToken) {...}`

Reply:
Same for `itemMetadata(...)` method btw :wink:

Reply:
Only the `LoadingCache` has a `getUnchecked` method,

Reply:
I see, then proceed to point 2 🙂

Reply:
Still WIP :wink: